src/ReachSets/ContinuousPost/BFFPS19/reach_blocks.jl

```diff
@@ -0,0 +1,301 @@
+                       n::Int,
+                       N::Union{Int, Nothing},
+                       output_function::Union{Function, Nothing},
+                       blocks::AbstractVector{Int},
+                       partition::AbstractVector{<:Union{AbstractVector{Int}, Int}},
+                       δ::NUM,
+                       guards_proj::Vector{<:LazySet{NUM}},
+                       block_options,
+                       vars::Vector{Int},
+                       termination::Function,
+                       progress_meter::Union{Progress, Nothing},
+                       res::Vector{<:ReachSet}
+                       )::Tuple{Int, Bool} where {NUM}
+    update!(progress_meter, 1)
+    array = CartesianProductArray(Xhat0[blocks])
+    X_store = (output_function == nothing) ?
+        array :
+        box_approximation(output_function(array))
+    t0 = zero(δ)
+    t1 = δ
+
+    diff_blocks = setdiff(collect(1:length(partition)),blocks)
+
+    b = length(blocks)
+    bd = length(diff_blocks)
+
+    array_d = CartesianProductArray(Xhat0[diff_blocks])
+    X_store_d = (output_function == nothing) ?
+            array_d :
+            box_approximation(output_function(array_d))
+    X_store = combine_cpas(X_store, X_store_d, blocks, diff_blocks)
+    store!(res, 1, X_store, t0, t1, N)
+    terminate, skip = termination(1, X_store, t0)
+    if terminate
+        return 1, skip
+    end
+
+    b = length(blocks)
+    Xhatk = Vector{LazySet{NUM}}(undef, b)
+    Xhatk_d = Vector{LazySet{NUM}}(undef, bd)
+    ϕpowerk = copy(ϕ)
+
+    if U != nothing
+        Whatk_blocks = Vector{LazySet{NUM}}(undef, b)
+        inputs = next_set(U)
+        @inbounds for i in 1:b
+            bi = partition[blocks[i]]
+            Whatk_blocks[i] = overapproximate_inputs(1, blocks[i], proj(bi, n) * inputs)
+       end
+    end
+
+    if U != nothing
+       Whatk_diff_blocks = Vector{LazySet{NUM}}(undef, bd)
+       inputs = next_set(U)
+       @inbounds for i in 1:bd
+           bi = partition[diff_blocks[i]]
+           Whatk_diff_blocks[i] = overapproximate_inputs(1, diff_blocks[i], proj(bi, n) * inputs)
+        end
+    end
+
+    k = 2
+    @inbounds while true
+        update!(progress_meter, k)
+        for i in 1:b
+            bi = partition[blocks[i]]
+            Xhatk_bi = ZeroSet(length(bi))
+            for (j, bj) in enumerate(partition)
+                block = ϕpowerk[bi, bj]
+                if !iszero(block)
+                    Xhatk_bi = Xhatk_bi + block * Xhat0[j]
+                end
+            end
+            Xhatk_bi_lazy = (U == nothing ? Xhatk_bi : Xhatk_bi + Whatk_blocks[i])
+            Xhatk[i] = (output_function == nothing) ?
+                overapproximate(blocks[i], Xhatk_bi_lazy) :
+                Xhatk_bi_lazy
+        end
+        array = CartesianProductArray(copy(Xhatk))
+        X_store = (output_function == nothing) ?
+            array :
+            box_approximation(output_function(array))
```

Comment:
Can you outsource everything starting from `for i in 1:b` to a helper function? Because you basically have the same code below, just with different indices. It would be something like:
```julia
@inline function deco_post_sparse(b, blocks, Whatk_blocks, Xhatk, partition,
                                  ϕpowerk, Xhat0, output_function)
     ... # copy code here
    return X_store
end
```